oak-blob-composite/src/main/java/org/apache/jackrabbit/oak/blob/composite/CompositeDataStoreService.java

```diff
@@ -0,0 +1,218 @@
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.ComponentContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jcr.RepositoryException;
+import java.io.Closeable;
+import java.util.Dictionary;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+import static org.apache.jackrabbit.oak.commons.IOUtils.closeQuietly;
+
+@Component(policy = ConfigurationPolicy.REQUIRE, name = CompositeDataStoreService.NAME)
+public class CompositeDataStoreService extends AbstractDataStoreService {
+    public static final String NAME = "org.apache.jackrabbit.oak.plugins.blob.datastore.CompositeDataStore";
+    private static Logger log = LoggerFactory.getLogger(CompositeDataStoreService.class);
+
+    private static final String DESCRIPTION = "oak.composite.datastore.description";
+
+    private CompositeDataStore dataStore = null;
+
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL_MULTIPLE,
+            policy = ReferencePolicy.DYNAMIC,
+            bind = "addDelegateDataStore",
+            unbind = "removeDelegateDataStore",
+            referenceInterface = DataStoreProvider.class,
+            target="(!(service.pid=org.apache.jackrabbit.oak.plugins.blob.datastore.CompositeDataStore))"
+    )
+    private List<DelegateDataStore> delegateDataStores = Lists.newArrayList();
+
+    private ComponentContext context;
+    private Map<String, Object> config = Maps.newConcurrentMap();
+
+    private boolean isServiceRegistered = false;
+    private Closer closer = Closer.create();
+
+    @Override
+    protected DataStore createDataStore(ComponentContext context, Map<String, Object> config) {
+        log.info("Creating Composite Data Store");
+
+        this.context = context;
+        if (null == this.config) {
+            this.config = config;
+        }
+        else {
+            for (Map.Entry<String, Object> entry : config.entrySet()) {
+                this.config.putIfAbsent(entry.getKey(), entry.getValue());
+            }
+        }
+
+        registerCompositeDataStore();
+
+        return dataStore;
+    }
+
+    private void registerCompositeDataStore() {
+        if (null == dataStore) {
+            Properties properties = null;
+            new Properties();
+            if (null != config) {
+                properties = new Properties();
+                properties.putAll(config);
+            }
+
+            if (null == properties) {
+                log.error("Composite Data Store configuration missing");
+                return;
+            }
+
+            Set<String> uniqueRoles = CompositeDataStore.getRolesFromConfig(properties);
+            if (0 == uniqueRoles.size()) {
+                log.error("No roles configured for Composite Data Store");
+                return;
+            }
+
+            if (delegateDataStores.size() != uniqueRoles.size()) {
```

Comment:
The check rather should be that all the required role-specific DataStore(s) have been initialized. This will ensure early and clear failures when configured incorrectly.

Reply:
I'm not sure what you have in mind.

One way I think this could be done easily would be to capture the role in addition to the delegate in `addDelegateDataStore()`.  If each role was put into a `Set` then the size of the added roles could be compared to the `uniqueRoles` and if they were the same then we would know that all configured roles had been added.

If you think we need to do something more involved I suppose we could also query the context for registered services, using `DataStoreProvider.class` as the service and using each role as the filter.

Do either of those match what you had in mind or are you thinking of something different?

Reply:
No what I had in mind was a simple check which determines that the all the roles configured have their respective delegates configured. This will just be a comparison of all the roles collected from the configured delegates (DelegateDataStore provides a method to return the role iirc) and the configured roles.